File: Find_First_and_Last_Position_of_Element.py
# Created by Aashish Adhikari at 2:58 PM 12/25/2020

'''
Time Complexity:
searchRange() : O(2 * logn) since two consecutive binary searches

Space Complexity:
searchRange(): O(1) since the auxiliary space created for each call of searchRange() is a constant
                Regardless of how many times the while loop executes, the same variable is reused.

Passes LettCode: Yes

Issue during implementation: No
'''

class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """

        l = 0
        r = len(nums)-1

        low = None
        high = None

        # Binary search to find the leftmost target
        while l <= r:

            mid = l + (r-l)//2


            if nums[mid] == target:

                if mid != 0:

                    if nums[mid-1] != target:
                        low = mid
                        break
                    else:
                        r = mid-1
                else:
                    low = 0
                    break
            elif nums[mid] < target:

                l = mid+1
            else:
                r = mid-1

        if low is None:
            # a single instance of target wasnt found, so return [-1,-1]
            return [-1, -1]

        # Binary search to find the rightmost target
        l = 0
        r = len(nums)-1
        while l <= r:

            mid = l + (r-l)//2


            if nums[mid] == target:

                if mid != len(nums)-1:

                    if nums[mid+1] != target:
                        high = mid
                        break
                    else:
                        l = mid+1
                else:
                    high = len(nums)-1
                    break
            elif nums[mid] < target:

                l = mid+1
            else:
                r = mid-1

        return [low, high]

File: FindFirstAndLastPositionOfElement.java
public class FindFirstAndLastPositionOfElement {

    // Time Complexity : O(log(N))
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No

    private int binarySearchFirst(int[] arr, int target, int low, int high) {
        while(low<=high) {
            int mid = (low+(high-low))/2;
            if(arr[mid] == target) {
                if(mid == 0 || arr[mid] > arr[mid-1]) {
                    return mid;
                } else {
                    high = mid -1; //keep moving left
                }
            } else if(arr[mid] > target){
                high = mid-1;
            } else {
                low = mid+1;
            }
        }
        return -1;
    }
    private int binarySearchLast(int[] arr, int target, int low, int high) {
        int n = arr.length;
        while(low<=high) {
            int mid = (low+(high-low))/2;
            if(arr[mid] == target) {
                if(mid == n-1 || arr[mid]<arr[mid+1] ) return mid;
                else {
                    low = mid+1;//keep moving right
                }
            } else if(arr[mid] > target) {
                high = mid-1;
            } else {
                low = mid+1;
            }
        }
        return -1;
    }
    public int[] searchRange(int[] nums, int target) {

        int n = nums.length;
        if(nums == null || nums.length == 0) return new int[]{-1,-1};
        if(nums[0] > target || nums[n-1] < target) return new int[]{-1,-1};

        int firstOccurance = binarySearchFirst(nums, target, 0, n-1);
        if(firstOccurance == -1) return new int[]{-1,-1};
        int lastOccurance = binarySearchLast(nums, target, firstOccurance, n-1);
        return new int[]{firstOccurance, lastOccurance};

    }
}
